### DISCUSSION TITLE
Celery configuration elements not shown to be fetched with `_CMD` pattern

### DISCUSSION THREAD

**Discussion by dkalamar** [apache/airflow#34030] [General]
### Apache Airflow version

2.7.0

### What happened

On the UI, certain celery configuration variables fail to show evaluated non-default value from when `_CMD` pattern is applied when specified in `Airflow.cfg`. These include:
```
[celery]
broker_url
result_backend
flower_basic_auth
```
Instead, the `_CMD` is listed as its own configuration variable and the displayed variable takes its value from `default`. The value is still properly applied, but not reflected in the UI.

### What you think should happen instead

Except from Airflow Configuration UI page:

Section | Key | Value | Source
-- | -- | -- | --
celery | broker_url | redis://redis:6379/0 | default-pre-2-7
celery | flower_basic_auth |   | default-pre-2-7
celery | result_backend_cmd | /usr/bin/env bash -c "echo -n db+postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}:${POSTGRES_PORT}/${POSTGRES_DB}" | airflow.cfg
celery | broker_url_cmd | /usr/bin/env bash -c "echo -n redis://${REDIS_USER}:${REDIS_PASSWORD}@${REDIS_HOST}:${REDIS_PORT}/0" | airflow.cfg
celery | flower_basic_auth_cmd | /usr/bin/env bash -c "echo -n ${FLOWER_USER}:${FLOWER_PASSWORD}" | airflow.cfg
database | sql_alchemy_conn | < hidden > | cmd

We would expect celery configuration variables to be shown to have been evaluated in the same way as `sql_alchemy_conn_cmd`. Instead, `broker_url` and `flower_basic_auth` display their default value. `result_backend` does not appear at all. Despite the UI bug, the values do appear to evaluate correctly in practice, as seen in this excerpt from `airflow config list`:
```
[database]
sql_alchemy_conn = postgresql://airflow:airflow@postgres:5432/airflow
sql_alchemy_conn_cmd = /usr/bin/env bash -c "echo -n postgresql://airflow:airflow@postgres:5432/airflow"

[celery]
broker_url = redis://:samplepassword@redis:6379/0
result_backend = db+postgresql://airflow:airflow@postgres:5432/airflow
flower_basic_auth = flower:flower
result_backend_cmd = /usr/bin/env bash -c "echo -n db+postgresql://airflow:airflow@postgres:5432/airflow"
broker_url_cmd = /usr/bin/env bash -c "echo -n redis://:samplepassword@redis:6379/0"
flower_basic_auth_cmd = /usr/bin/env bash -c "echo -n flower:flower"
```

### How to reproduce
Set environment variables:
```
export POSTGRES_USER=airflow
export POSTGRES_PASSWORD=airflow
export POSTGRES_HOST=postgres
export POSTGRES_PORT=5432
export POSTGRES_DB=postgres

export REDIS_PASSWORD=samplepassword
export REDIS_HOST=redis
export REDIS_PORT=6739
```
Airflow.cfg excerpt:
```
[celery]
result_backend_cmd = /usr/bin/env bash -c "echo -n db+postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}:${POSTGRES_PORT}/${POSTGRES_DB}"
broker_url_cmd = /usr/bin/env bash -c "echo -n redis://${REDIS_USER}:${REDIS_PASSWORD}@${REDIS_HOST}:${REDIS_PORT}/0"

[database]
sql_alchemy_conn_cmd = /usr/bin/env bash -c "echo -n postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}:${POSTGRES_PORT}/${POSTGRES_DB}"
```

### Operating System

ubuntu22.04 

### Versions of Apache Airflow Providers

Package Name | Version | Description
-- | -- | --
apache-airflow-providers-amazon | 8.5.1 | Amazon integration (including Amazon Web Services (AWS)).<br>
apache-airflow-providers-celery | 3.3.3 | Celery<br>
apache-airflow-providers-common-sql | 1.7.0 | Common SQL Provider<br>
apache-airflow-providers-ftp | 3.5.0 | File Transfer Protocol (FTP)<br>
apache-airflow-providers-http | 4.5.0 | Hypertext Transfer Protocol (HTTP)<br>
apache-airflow-providers-imap | 3.3.0 | Internet Message Access Protocol (IMAP)<br>
apache-airflow-providers-pagerduty | 3.3.0 | Pagerduty<br>
apache-airflow-providers-postgres | 5.6.0 | PostgreSQL<br>
apache-airflow-providers-redis | 3.3.1 | Redis<br>
apache-airflow-providers-slack | 7.3.2 | Slack<br>
apache-airflow-providers-snowflake | 4.4.2 | Snowflake<br>
apache-airflow-providers-sqlite | 3.4.3 | SQLite<br>
apache-airflow-providers-ssh | 3.7.1 | Secure Shell (SSH)<br>

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
I could not reproduce it.

If it was for celery only - then you probably do not have latest celey provider installed. As of Airlfow 2.7.0, configuration coming from providers comes from the providers. In order to use and configure celery provider, you need to have celery provider installed.

So it could be that you have no celery provider installed.


I confifgured it this way:

```
[core]
fernet_key_cmd = /usr/bin/env bash -c "echo -n XXX"

[celery]
broker_url_cmd = /usr/bin/env bash -c "echo -n ZZZ"
```

And it seems to be working fine:

```
root@44408d006bc4:/opt/airflow# airflow config get-value core fernet_key
XXX
root@44408d006bc4:/opt/airflow# airflow config get-value celery broker_url
The option [celery/broker_url] is not found in config.
root@44408d006bc4:/opt/airflow# pip install apache-airflow-providers-celery
Collecting apache-airflow-providers-celery
  Obtaining dependency information for apache-airflow-providers-celery from https://files.pythonhosted.org/packages/ae/37/6751f74137201740735e93e77365895fefe3fab80f4a5ec71cb62964c013/apache_airflow_providers_celery-3.3.3-py3-none-any.whl.metadata
  Downloading apache_airflow_providers_celery-3.3.3-py3-none-any.whl.metadata (5.8 kB)
.....
root@44408d006bc4:/opt/airflow# airflow config get-value celery broker_url
ZZZ

```



But with fernet_key - notihng changed in Airflow 2.7.0 - so It's surprising it does not work for you. Possibly you have some configuration that overrides it? 

You need dig deeper - where you do not see it, how did you check it , what was the command you run so that it did not work etc - because it does not seem reproducible.

**Comment by potiuk** [comment]
> Instead the _CMD is listed as its own configuration variable and the evaluated variable takes its value from default.

Also I do not understand what do you mean by `listed` ? where ?

**Comment by dkalamar** [comment]
@potiuk

I was able to dive in further, the failed evaluation I saw on `FERNET_KEY` and `FLOWER_BASIC_AUTH` appear to be unrelated configuration errors on my end.

There is, however, still an issue with the way Celery Configuration's are displayed on the UI, despite being properly evaluated. I am running the most recent provider package version `apache-airflow-providers-celery == 3.3.3`

**Comment by dkalamar** [comment]
It appears that celery values are skipped from evaluation in `AirflowConfigParser._get_option_from_commands` since `include_providers` is set to false on [airflow/configuration:L204](https://github.com/apache/airflow/blob/3ae6b4e86fe807c00bd736c59df58733df2b9bf9/airflow/configuration.py#L204C42-L204C42)
```
self.configuration_description = retrieve_configuration_description(include_providers=False)
```

I would expect `ProvidersManager().initialize_providers_configuration()` to be called by default.

**Comment by potiuk** [comment]
What configuration in the UI  ?

**Comment by potiuk** [comment]
Screenshots?

**Comment by potiuk** [comment]
The  `self.configuration_description = retrieve_configuration_description(include_providers=False)` is deliberately set - and mainly because our configuration is needed often when commands are displayed and we cannot load provider configuration then. Loading provider's configuration is somewhat costly, because we need to find all providers installed, retrieve their `get_provider_info` entrypoint and update configuration with them.

Generally speaking loading configuration now is done in two stages:

1) config parser is loaded first without providers (as you rightfully noticed
2) config parser is UPDATED with provider configuration information after the configuration is discovered from provoders 

https://github.com/apache/airflow/blob/3ae6b4e86fe807c00bd736c59df58733df2b9bf9/airflow/providers_manager.py#L548

When `initialize_providers_configuration` is called on ProvidersManager, it retrieves all the provider information. and once this information is retrieved it runs ` conf.load_providers_configuration()` that updates information about provider's configuraiton.

I believe all the places that airflow needs, the step of initializing provider's configuration should be done before. But maybe there is a place where it is not - but you need to tell me which place it is.

If this is some of your code that you wrote - you need to update your code to also load provider configuration.

Converting it into discussion - because we seem to be discussing about some internals of Airflow and you are talking about somethign rather custom.

**Comment by kakashif** [comment]
We are running into the same situation with 2.7.1 upgrade, from 2.6.0.  We have the same setup as far as the OS and the deployment goes but here is the provider list:
```
apache-airflow[pandas]==2.7.1
apache-airflow-providers-amazon
apache-airflow-providers-celery
apache-airflow-providers-hashicorp
apache-airflow-providers-postgres
apache-airflow-providers-redis
apache-airflow-providers-slack
apache-airflow-providers-ssh
apache-airflow-providers-snowflake
apache-airflow-providers-github
```

We have a different way of accessing our vault keys so we have a custom backend provider setup like so:
```
class CustomVaultProvider(BaseSecretsBackend, LoggingMixin):
```
And the env vars are correctly setup:
```
AIRFLOW__CELERY__BROKER_URL_SECRET: custom_airflow/broker_url
AIRFLOW__CELERY__RESULT_BACKEND_SECRET: custom_airflow/result_backend
AIRFLOW__CELERY__SSL_ACTIVE: True
AIRFLOW__CORE__FERNET_KEY_SECRET: custom_airflow/fernet_key
AIRFLOW__SECRETS__BACKEND: backends.CustomVaultProvider
```

We also have an http listener added as a plugin as we are required to have a health check for the worker like so:
```
class CeleryWorkerHTTPHealthCheckPlugin(AirflowPlugin):
```

When deploying after the upgrade, I see the following debug statement in the log: 
```
++ airflow celery worker
[2023-09-23T01:16:17.123+0000] {settings.py:212} DEBUG - Setting up DB connection pool (PID 19)
[2023-09-23T01:16:17.124+0000] {settings.py:317} DEBUG - settings.prepare_engine_args(): Using pool settings. pool_size=5, max_overflow=10, pool_recycle=1800, pid=19
[2023-09-23T01:16:17.237+0000] {plugins_manager.py:328} DEBUG - Loading plugins
[2023-09-23T01:16:17.237+0000] {plugins_manager.py:252} DEBUG - Loading plugins from directory: /usr/local/airflow/plugins
[2023-09-23T01:16:17.238+0000] {plugins_manager.py:267} DEBUG - Importing plugin module /usr/local/airflow/plugins/envar_secrets_first.py
[2023-09-23T01:16:17.238+0000] {envar_secrets_first.py:42} INFO - Enabled use_environment_variables_first plugin. (a custom plugin we wrote)
[2023-09-23T01:16:17.254+0000] {CustomVaultProvider.py:114} INFO - About to read secret CONFIG/custom_airflow/sql_alchemy_conn with vault id #####
...
...
...
[2023-09-23T01:16:18.277+0000] {plugins_manager.py:267} DEBUG - Importing plugin module /usr/local/airflow/plugins/scheduler_http_health.py
[2023-09-23T01:16:18.277+0000] {scheduler_http_health.py:46} INFO - Wrapping scheduler command with http server
[2023-09-23T01:16:18.491+0000] {default_celery.py:68} DEBUG - Value for celery result_backend not found. Using sql_alchemy_conn with db+ prefix.
[2023-09-23T01:16:18.491+0000] {CustomVaultProvider.py:114} INFO - About to read secret CONFIG/custom_airflow/sql_alchemy_conn with vault id #####
```

After seeing this post, I added the following call inside the first plugin I saw in the log files that got loaded, an env related plugin, envar_secrets_first.py, which can bee seen in the log entries above:
```
from airflow.providers_manager import ProvidersManager

class UseEnvironmentVariablesFirstPlugin(AirflowPlugin):

    # load the configuration for all providers.  Not sure how this is going to work with Airflow as this is a plugin
    # inception.  We are forcing to load the configuration inside a plugin.
    # todo: verify if this is needed in other services too

    ProvidersManager().initialize_providers_configuration()
```

Even after doing this, I still see a the following debug statements and the worker starts with local redis connection in the log files:
```

[2023-09-24T17:31:09.365+0000] {default_celery.py:68} DEBUG - Value for celery result_backend not found. Using sql_alchemy_conn with db+ prefix.

 -------------- celery@a952c9e82c36 v5.3.4 (emerald-rush)
--- ***** ----- 
-- ******* ---- Linux-MASKED
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         airflow.executors.celery_executor:0x7fbb5144dc40
- ** ---------- .> transport:   redis://redis:6379/0
- ** ---------- .> results:     postgresql+psycopg2://MASKED
- *** --- * --- .> concurrency: 16 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> default          exchange=default(direct) key=default
```


tldr; basically what I am trying to get to is that for the airflow backend or any other secrets/connections (fernet key, sql alchemy conn, sso integration secrets), it correctly uses my custom backend but it does not uses the secret backend for loading the celery broker URL or the celery result backend.  Other than the few plugins (worker health, custom scheduler health, env var first and the vault backend), there are no other plugins.  And I am using the following constraint file: https://github.com/apache/airflow/blob/constraints-2-7/constraints-3.9.txt 

FYI, I have not tried the `_cmd` setup.  My coworker use `_cmd` and for them it seems to be working but of course they have a completely different setup and are on 2.7.0 whereas I need to be on 2.7.1 due to the SSO bug that has been fixed in 2.7.1.

**Reply by kakashif** [reply]
If it helps, this issue is not in the scheduler at least. I have not deployed the web server with the debug logs enabled but at least in the scheduler, I see it reach back to the custom vault backend for the broker URL and the result backend as I have added a log statement for every secret it reads.  But I never see the log statements in the worker logs for the two settings.

**Reply by kakashif** [reply]
Okay moving it around and finally got it to work.  
I have seen the coworkers add it to the http plugin they wrote to create the HTTP health checks for the worker, it is a company requirement so we just have the celery ping call wrapped in an endpoint.

Do you have a particular suggestion or can point me to a docs where the `ProvidersManager().initialize_providers_configuration()`  call should live?
